### DISCUSSION TITLE
Column is null after value change

### DISCUSSION THREAD

**Discussion by AlfaGTA** [Mimetis/Dotmim.Sync#609] [Q&A]
I'm trying to use the sample Xamaring.Forms project to sync between my remote db on SQL and an Android emulator.
The first synchronization is successful and 45k records are updated. However, if I change any value of one of the synchronized tables, and resynchronize, I get the following exception:

Column is null
-----------------------
   at Dotmim.Sync.BaseOrchestrator.RaiseError(Exception exception) in D:\Progetti\Vari\Dotmim.Sync-master\Dotmim.Sync-master\Projects\Dotmim.Sync.Core\Orchestrators\BaseOrchestrator.cs:line 250
   at Dotmim.Sync.RemoteOrchestrator.ApplyThenGetChangesAsync(ScopeInfo clientScope, BatchInfo clientBatchInfo, CancellationToken cancellationToken, IProgress`1 progress) in D:\Progetti\Vari\Dotmim.Sync-master\Dotmim.Sync-master\Projects\Dotmim.Sync.Core\Orchestrators\RemoteOrchestrator.cs:line 278
   at Dotmim.Sync.RemoteOrchestrator.ApplyThenGetChangesAsync(ScopeInfo clientScope, BatchInfo clientBatchInfo, CancellationToken cancellationToken, IProgress`1 progress) in D:\Progetti\Vari\Dotmim.Sync-master\Dotmim.Sync-master\Projects\Dotmim.Sync.Core\Orchestrators\RemoteOrchestrator.cs:line 283
   at Dotmim.Sync.Web.Server.WebServerOrchestrator.ApplyThenGetChangesAsync2(HttpContext httpContext, HttpMessageSendChangesRequest httpMessage, SessionCache sessionCache, Int32 clientBatchSize, CancellationToken cancellationToken, IProgress`1 progress) in D:\Progetti\Vari\Dotmim.Sync-master\Dotmim.Sync-master\Projects\Dotmim.Sync.Web.Server\WebServerOrchestrator.cs:line 735
   at Dotmim.Sync.Web.Server.WebServerOrchestrator.HandleRequestAsync(HttpContext httpContext, Action`1 action, CancellationToken cancellationToken, IProgress`1 progress) in D:\Progetti\Vari\Dotmim.Sync-master\Dotmim.Sync-master\Projects\Dotmim.Sync.Web.Server\WebServerOrchestrator.cs:line 174
-----------------------
-----------------------
INNER EXCEPTION
-----------------------
Column is null
-----------------------
   at Dotmim.Sync.SyncRow.set_Item(String columnName, Object value) in D:\Progetti\Vari\Dotmim.Sync-master\Dotmim.Sync-master\Projects\Dotmim.Sync.Core\Set\SyncRow.cs:line 115
   at Dotmim.Sync.BaseOrchestrator.CreateSyncRowFromReader(IDataReader dataReader, SyncTable table) in D:\Progetti\Vari\Dotmim.Sync-master\Dotmim.Sync-master\Projects\Dotmim.Sync.Core\Orchestrators\BaseOrchestrator.GetChanges.cs:line 416
   at Dotmim.Sync.BaseOrchestrator.InternalGetChangesAsync(SyncContext context, MessageGetChangesBatch message, DbConnection connection, DbTransaction transaction, CancellationToken cancellationToken, IProgress`1 progress) in D:\Progetti\Vari\Dotmim.Sync-master\Dotmim.Sync-master\Projects\Dotmim.Sync.Core\Orchestrators\BaseOrchestrator.GetChanges.cs:line 113
   at Dotmim.Sync.RemoteOrchestrator.ApplyThenGetChangesAsync(ScopeInfo clientScope, BatchInfo clientBatchInfo, CancellationToken cancellationToken, IProgress`1 progress) in D:\Progetti\Vari\Dotmim.Sync-master\Dotmim.Sync-master\Projects\Dotmim.Sync.Core\Orchestrators\RemoteOrchestrator.cs:line 243
-----------------------

**Comment by Mimetis** [comment]
Can you share your code ? especially the configuration of your sync, with tables and columns (if any are specified) ?
I think you have made a filter on columns that allow null on client, but not on the server ....

**Reply by AlfaGTA** [reply]
This is the server side code, in Startup.cs:

```
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddControllers();

            // [Required]: Handling multiple sessions
            services.AddDistributedMemoryCache();
            services.AddSession(options => options.IdleTimeout = TimeSpan.FromMinutes(30));

            // [Required]: Get a connection string to your server data source
            //var connectionString = Configuration.GetSection("ConnectionStrings")["SqlConnection"];
            string connString = $@"Server=UTENTE-PC;Database=Vendo;User Id=sa;Password=***;";

            // Sync options
            var options = new SyncOptions
            {
                SnapshotsDirectory = Path.Combine(SyncOptions.GetDefaultUserBatchDiretory(), "Snapshots"),
                BatchSize = 2000,
            };

            //// [Required] Tables involved in the sync process:
            //var tables = new string[] {"ProductCategory", "ProductModel", "Product",
            //"Address", "Customer", "CustomerAddress", "SalesOrderHeader", "SalesOrderDetail" };
            var tables = new string[] { "Articoli", "DatiStorico", "Magazzini" };

            var setup = new SyncSetup(tables);

            // add a SqlSyncProvider acting as the server hub
            services.AddSyncServer<SqlSyncProvider>(connString, setup, options);
        }
```

As you can see, the code is pretty similar to the sample code provided by you.
I haven't applied any filters on that tables. 
Here is the client side code on Xamarin.Forms that throw the exception:

```
               var progress = new SynchronousProgress<ProgressArgs>(args =>
              {
                  Device.BeginInvokeOnMainThread(() =>
                  {
                      this.SyncProgress = args.ProgressPercentage;
                      this.SyncProgressionText = args.Message;
                  });
              });

              var r = await this.syncAgent.SynchronizeAsync(syncType, progress);   // <--- Normal sync type

              Device.BeginInvokeOnMainThread(() => this.SyncProgressionText = r.ToString());
```

**Reply by Mimetis** [reply]
Can you share your schema (just the 3 tables you are working on)

**Reply by Mimetis** [reply]
Can you try without the snapshot option as well ?

**Comment by AlfaGTA** [comment]
I removed shapshot option, but nothing.
I add a detail: when I try to resync, the process retry sending http request up to 10 times and after throw the exception in object (I don't know if it can be a useful detail).

Here is the schema of one of these table:
CREATE TABLE [dbo].[DatiStorico] (
	[CodCli] [nvarchar] (7) NOT NULL ,
	[Agente] [int] NOT NULL ,
	[Data] [datetime] NOT NULL ,
	[CodArt] [nvarchar] (15) NOT NULL ,
	[Tipo] [nvarchar] (2) NOT NULL ,
	[Qta] [float] NOT NULL ,
	[UltimoPrezzo] [float] NOT NULL ,
	[IDData] [int] NOT NULL ,
	[ID] [int] IDENTITY (1, 1) NOT NULL ,
	[Sconti] [nvarchar] (15) NOT NULL ,
	[Descrizione] [nvarchar] (255) NULL
	CONSTRAINT [PK_DatiStorico] PRIMARY KEY   
	(
		[ID]
	))

CREATE NONCLUSTERED INDEX [IX_DatiStorico] ON [dbo].[DatiStorico]
([CodCli] ASC, [CodArt] ASC, [Tipo] ASC)

CREATE NONCLUSTERED INDEX [IX_CodArt] ON [dbo].[DatiStorico]
([CodArt] ASC)

CREATE NONCLUSTERED INDEX [IX_CodCli] ON [dbo].[DatiStorico]
([CodCli] ASC)

**Reply by Mimetis** [reply]
Seems you have 3 tables, but you shared only 1

**Reply by AlfaGTA** [reply]
CREATE TABLE [dbo].[Articoli] (
[CodArt] [nvarchar] (15) NOT NULL ,
[Descrizione] [nvarchar] (140) NOT NULL ,
[UM] [nvarchar] (2) NULL ,
[Barcode] [nvarchar] (6) NULL ,
[CodGrup] [nvarchar] (7) NOT NULL ,
[CodAli] [nvarchar] (10) NULL ,
[UltimoCosto] [float] NOT NULL ,
[MoltUM2] [float] NULL ,
[UM2] [nvarchar] (2) NULL ,
[MoltCF] [float] NULL ,
[TipoArt] [tinyint] NOT NULL ,
[Scaglione] [nvarchar] (1) NULL ,
[ID] [int] IDENTITY (1, 1) NOT NULL
CONSTRAINT [PK_Articoli] PRIMARY KEY
([ID]))

CREATE NONCLUSTERED INDEX [IX_CodGrup] ON [dbo].[Articoli]
([CodGrup] ASC)

CREATE UNIQUE NONCLUSTERED INDEX [IX_Codice] ON [dbo].[Articoli]
([CodArt] ASC)

CREATE TABLE [dbo].[Magazzini] (
[CodArt] [nvarchar] (15) NOT NULL ,
[Esistenza] [float] NOT NULL ,
[CodMaga] [nvarchar] (2) NOT NULL ,
[Agente] [int] NULL ,
[Esporta] [bit] NOT NULL ,
[ID] [int] IDENTITY (1, 1) NOT NULL
CONSTRAINT [PK_Magazzini] PRIMARY KEY
([ID]))

CREATE UNIQUE NONCLUSTERED INDEX [IX_Codice] ON [dbo].[Magazzini]
([CodArt] ASC, [CodMaga] ASC)

CREATE NONCLUSTERED INDEX [IX_CodArt] ON [dbo].[Magazzini]
([CodArt] ASC)

**Reply by AlfaGTA** [reply]
In sync configuration I replaced SqlSyncProvider with SqlSyncChangeTrackingProvider, like so:

```
            //services.AddSyncServer<SqlSyncProvider>(connString, setup, options);
            services.AddSyncServer<SqlSyncChangeTrackingProvider>(connString, setup, options);
```

And now, the exception no longer occurs, but I am not getting the changes.

**Reply by Mimetis** [reply]
I don't have the time nowadays, but I will take a look later this week

**Reply by AlfaGTA** [reply]
After a deep investigation, I found that the error is throwned in Dotmim.Sync.Core\Set\SyncRow.cs:line 115: The column "update_scope_id" is not founded, but this column exists in DatiStorico_tracking table, not in DatiStorico table.

**Reply by AlfaGTA** [reply]
In Dotmim.Sync.Core\Orchestrators\BaseOrchestrator.GetChanges.cs, CreateSyncRowFromReader method, I replaced original code with following:

```
                // >>> original code
                //if (columnName == "sync_update_scope_id")  
                //    continue;
               // <<<

                if (columnName == "sync_update_scope_id" || columnName == "update_scope_id")
                    continue;
```

It looks like the column update_scope_id needs to be skipped. 
Now it works perfectly, but I don't know if it is the best solution.

**Reply by Mimetis** [reply]
Did you upgrade from `v0.8.0 ` to `v0.8.1` ? (and the bug appears after migrating to `v0.8.1` )

If yes, you need to apply the migration on your server side:

``` cs
// Upgrade to last version
if (await remoteOrchestrator.NeedsToUpgradeAsync())
   await remoteOrchestrator.UpgradeAsync(progress:progress);
```

You can check before and after the migration in any  `TABLE_changes` stored procedure:

BEFORE (`v0.8.0`)
``` sql
SELECT DISTINCT	[side].[ProductCategoryID], 
	[base].[ParentProductCategoryID], 
	[base].[Name], 
	[base].[rowguid], 
	[base].[ModifiedDate], 
	[side].[sync_row_is_tombstone], 
	[side].[update_scope_id]
```

AFTER (`v0.8.1`)
``` sql
SELECT DISTINCT	[side].[ProductCategoryID], 
	[base].[ParentProductCategoryID], 
	[base].[Name], 
	[base].[rowguid], 
	[base].[ModifiedDate], 
	[side].[sync_row_is_tombstone] as [sync_row_is_tombstone], 
	[side].[update_scope_id] as [sync_update_scope_id]
```
